File: src/linear_algebra_vector_spaces_medium.py
"""
linear_algebra - vector_spaces (medium)
Generated: 2026-02-22T05:53:56.640327
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Linear independence check for 2-3 vectors in R^2 or R^3
        dim = choice([2, 3])
        num_vecs = choice([2, 3])
        
        if randint(0, 1) == 0:  # Make linearly dependent
            if dim == 2:
                v1_vals = [nonzero(-4, 4), nonzero(-4, 4)]
                scalar = choice([2, 3, -2, -3])
                v2_vals = [scalar * v1_vals[0], scalar * v1_vals[1]]
                vectors = [Matrix(v1_vals), Matrix(v2_vals)]
                ans = False
            else:  # dim == 3
                v1_vals = [nonzero(-3, 3), nonzero(-3, 3), nonzero(-3, 3)]
                v2_vals = [nonzero(-3, 3), nonzero(-3, 3), nonzero(-3, 3)]
                # Make v3 a linear combination
                s1, s2 = choice([1, 2, -1]), choice([1, 2, -1])
                v3_vals = [s1*v1_vals[0] + s2*v2_vals[0], 
                          s1*v1_vals[1] + s2*v2_vals[1],
                          s1*v1_vals[2] + s2*v2_vals[2]]
                vectors = [Matrix(v1_vals), Matrix(v2_vals), Matrix(v3_vals)]
                ans = False
        else:  # Make linearly independent
            if dim == 2 and num_vecs == 2:
                v1_vals = [nonzero(-4, 4), nonzero(-4, 4)]
                v2_vals = [nonzero(-4, 4), nonzero(-4, 4)]
                # Ensure not parallel
                while v1_vals[0] * v2_vals[1] == v1_vals[1] * v2_vals[0]:
                    v2_vals = [nonzero(-4, 4), nonzero(-4, 4)]
                vectors = [Matrix(v1_vals), Matrix(v2_vals)]
                ans = True
            else:  # 3D
                v1_vals = [1, 0, 0]
                v2_vals = [0, 1, 0]
                if num_vecs == 3:
                    v3_vals = [0, 0, 1]
                    vectors = [Matrix(v1_vals), Matrix(v2_vals), Matrix(v3_vals)]
                else:
                    vectors = [Matrix(v1_vals), Matrix(v2_vals)]
                ans = True
        
        vec_strs = [f"\\begin{{bmatrix}} {' \\\\ '.join(map(str, v))} \\end{{bmatrix}}" 
                    for v in vectors]
        question = f"Are the vectors {', '.join(vec_strs)} linearly independent?"
        
        M = Matrix.hstack(*vectors)
        rref_M = M.rref()[0]
        
        solution = steps(
            f"Form matrix with vectors as columns: ${latex(M)}$",
            f"Reduce to RREF: ${latex(rref_M)}$",
            f"{'All columns have pivots, so vectors are linearly independent.' if ans else 'Not all columns have pivots, so vectors are linearly dependent.'}"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1300, 1450),
            topic="linear_algebra/vector_spaces",
            solution=solution
        )
    
    elif problem_type == 2:
        # Find basis for span of vectors
        dim = choice([2, 3])
        
        if dim == 2:
            v1_vals = [nonzero(-3, 3), nonzero(-3, 3)]
            scalar = choice([2, 3, -2])
            v2_vals = [scalar * v1_vals[0], scalar * v1_vals[1]]
            v3_vals = [nonzero(-3, 3), nonzero(-3, 3)]
            while v3_vals[0] * v1_vals[1] == v3_vals[1] * v1_vals[0]:
                v3_vals = [nonzero(-3, 3), nonzero(-3, 3)]
            
            vectors = [Matrix(v1_vals), Matrix(v2_vals), Matrix(v3_vals)]
            basis = [Matrix(v1_vals), Matrix(v3_vals)]
        else:
            v1_vals = [1, 0, 0]
            v2_vals = [0, 1, 0]
            v3_vals = [1, 1, 0]
            vectors = [Matrix(v1_vals), Matrix(v2_vals), Matrix(v3_vals)]
            basis = [Matrix(v1_vals), Matrix(v2_vals)]
        
        vec_strs = [f"\\begin{{bmatrix}} {' \\\\ '.join(map(str, v))} \\end{{bmatrix}}" 
                    for v in vectors]
        question = f"Find a basis for the span of {', '.join(vec_strs)}. Enter your answer as a list of column vectors."
        
        ans = basis
        
        M = Matrix.hstack(*vectors)
        rref_M = M.rref()[0]
        
        solution = steps(
            f"Form matrix with vectors as columns: ${latex(M)}$",
            f"Reduce to RREF: ${latex(rref_M)}$",
            f"Pivot columns correspond to basis vectors",
            f"Basis: {', '.join([latex(v) for v in basis])}"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1400, 1550),
            topic="linear_algebra/vector_spaces",
            solution=solution
        )
    
    elif problem_type == 3:
        # Determine dimension of subspace
        dim = choice([3, 4])
        
        if dim == 3:
            # Span of 2 independent vectors in R^3
            v1_vals = [1, 0, nonzero(-2, 2)]
            v2_vals = [0, 1, nonzero(-2, 2)]
            ans = 2
            vectors = [Matrix(v1_vals), Matrix(v2_vals)]
        else:
            # Span of 3 independent vectors in R^4
            v1_vals = [1, 0, 0, 0]
            v2_vals = [0, 1, 0, 0]
            v3_vals = [0, 0, 1, 0]
            ans = 3
            vectors = [Matrix(v1_vals), Matrix(v2_vals), Matrix(v3_vals)]
        
        vec_strs = [f"\\begin{{bmatrix}} {' \\\\ '.join(map(str, v))} \\end{{bmatrix}}" 
                    for v in vectors]
        question = f"What is the dimension of the subspace spanned by {', '.join(vec_strs)}?"
        
        M = Matrix.hstack(*vectors)
        
        solution = steps(
            f"Form matrix with vectors as columns: ${latex(M)}$",
            f"Count the number of linearly independent vectors",
            f"The dimension is ${ans}$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1300, 1450),
            topic="linear_algebra/vector_spaces",
            solution=solution
        )
    
    elif problem_type == 4:
        # Check if vector is in span
        v1_vals = [1, 0, nonzero(-2, 2)]
        v2_vals = [0, 1, nonzero(-2, 2)]
        
        if randint(0, 1) == 0:  # In span
            c1, c2 = nonzero(-3, 3), nonzero(-3, 3)
            test_vals = [c1 * v1_vals[0] + c2 * v2_vals[0],
                        c1 * v1_vals[1] + c2 * v2_vals[1],
                        c1 * v1_vals[2] + c2 * v2_vals[2]]
            ans = True
        else:  # Not in span
            test_vals = [nonzero(-3, 3), nonzero(-3, 3), nonzero(-3, 3)]
            # Make sure third component doesn't match
            while test_vals[2] == test_vals[0] * v1_vals[2] + test_vals[1] * v2_vals[2]:
                test_vals[2] = nonzero(-3, 3)
            ans = False
        
        v1 = Matrix(v1_vals)
        v2 = Matrix(v2_vals)
        test = Matrix(test_vals)
        
        vec_strs = [f"\\begin{{bmatrix}} {' \\\\ '.join(map(str, v))} \\end{{bmatrix}}" 
                    for v in [v1, v2]]
        test_str = f"\\begin{{bmatrix}} {' \\\\ '.join(map(str, test))} \\end{{bmatrix}}"
        
        question = f"Is the vector {test_str} in the span of {', '.join(vec_strs)}?"
        
        M = Matrix.hstack(v1, v2, test)
        rref_M = M.rref()[0]
        
        solution = steps(
            f"Form augmented matrix: ${latex(M)}$",
            f"Reduce to RREF: ${latex(rref_M)}$",
            f"{'The system is consistent, so the vector is in the span.' if ans else 'The system is inconsistent, so the vector is not in the span.'}"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1350, 1500),
            topic="linear_algebra/vector_spaces",
            solution=solution
        )
    
    else:  # problem_type == 5
        # Find coordinates relative to a basis
        # Simple basis in R^2
        b1_vals = [1, nonzero(-2, 2)]
        b2_vals = [0, 1]
        
        c1, c2 = nonzero(-3, 3), nonzero(-3, 3)
        v_vals = [c1 * b1_vals[0] + c2 * b2_vals[0],
                 c1 * b1_vals[1] + c2 * b2_vals[1]]
        
        b1 = Matrix(b1_vals)
        b2 = Matrix(b2_vals)
        v = Matrix(v_vals)
        
        ans = Matrix([c1, c2])
        
        basis_strs = [f"\\begin{{bmatrix}} {' \\\\ '.join(map(str, vec))} \\end{{bmatrix}}" 
                     for vec in [b1, b2]]
        v_str = f"\\begin{{bmatrix}} {' \\\\ '.join(map(str, v))} \\end{{bmatrix}}"
        
        question = f"Find the coordinates of {v_str} relative to the basis $\\mathcal{{B}} = \\{{ {', '.join(basis_strs)} \\}}$. Enter as a column vector."
        
        P = Matrix.hstack(b1, b2)
        
        solution = steps(
            f"We need to solve ${latex(P)} \\mathbf{{c}} = {latex(v)}$",
            f"where $\\mathbf{{c}} = \\begin{{bmatrix}} c_1 \\\\ c_2 \\end{{bmatrix}}$",
            f"Solving gives $\\mathbf{{c}} = {latex(ans)}$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1400, 1600),
            topic="linear_algebra/vector_spaces",
            solution=solution
        )

emit(generate())

File: src/linear_algebra_vector_spaces_medium.jl
# linear_algebra - vector_spaces (medium)
# Generated: 2026-03-08T21:10:29.591772

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y z begin
    set_topic!("linear_algebra/vector_spaces")
    
    problem_type = rand(1:6)
    
    if problem_type == 1
        # Linear independence check (1200-1400)
        a1, a2, a3 = nonzero(-8, 8), nonzero(-8, 8), nonzero(-8, 8)
        b1, b2, b3 = randint(-8, 8), randint(-8, 8), randint(-8, 8)
        
        # Make third vector a linear combination of first two
        k1, k2 = nonzero(-3, 3), nonzero(-3, 3)
        c1 = k1 * a1 + k2 * b1
        c2 = k1 * a2 + k2 * b2
        c3 = k1 * a3 + k2 * b3
        
        v1 = "\\begin{bmatrix} $a1 \\\\ $a2 \\\\ $a3 \\end{bmatrix}"
        v2 = "\\begin{bmatrix} $b1 \\\\ $b2 \\\\ $b3 \\end{bmatrix}"
        v3 = "\\begin{bmatrix} $c1 \\\\ $c2 \\\\ $c3 \\end{bmatrix}"
        
        problem(
            question="Determine whether the vectors \$\\mathbf{v}_1 = $v1\$, \$\\mathbf{v}_2 = $v2\$, and \$\\mathbf{v}_3 = $v3\$ are linearly independent or linearly dependent.",
            answer="linearly dependent",
            difficulty=(1200, 1400),
            answer_type="choice",
            solution=steps(
                "Set up equation \$c_1\\mathbf{v}_1 + c_2\\mathbf{v}_2 + c_3\\mathbf{v}_3 = \\mathbf{0}\$",
                "Notice that \$\\mathbf{v}_3 = $(k1)\\mathbf{v}_1 + $(k2)\\mathbf{v}_2\$",
                "Since a nontrivial linear combination equals zero, the vectors are linearly dependent."
            ),
            time=120
        )
        
    elseif problem_type == 2
        # Basis verification (1300-1500)
        # Generate 2 vectors in R^2 that form a basis
        a1, a2 = nonzero(-7, 7), nonzero(-7, 7)
        b1, b2 = nonzero(-7, 7), nonzero(-7, 7)
        
        # Ensure they're linearly independent (det != 0)
        while a1 * b2 - a2 * b1 == 0
            b1, b2 = nonzero(-7, 7), nonzero(-7, 7)
        end
        
        v1 = "\\begin{bmatrix} $a1 \\\\ $a2 \\end{bmatrix}"
        v2 = "\\begin{bmatrix} $b1 \\\\ $b2 \\end{bmatrix}"
        
        problem(
            question="Verify that \$\\mathbf{v}_1 = $v1\$ and \$\\mathbf{v}_2 = $v2\$ form a basis for \$\\mathbb{R}^2\$. Is this a valid basis? Answer 'yes' or 'no'.",
            answer="yes",
            difficulty=(1300, 1500),
            answer_type="choice",
            solution=steps(
                "Check linear independence by computing determinant",
                sol("det", a1 * b2 - a2 * b1),
                "Since det ≠ 0, vectors are linearly independent. Two linearly independent vectors in \$\\mathbb{R}^2\$ form a basis."
            ),
            time=90
        )
        
    elseif problem_type == 3
        # Span verification (1200-1400)
        a1, a2 = nonzero(-6, 6), nonzero(-6, 6)
        k = nonzero(-4, 4)
        b1, b2 = k * a1, k * a2
        
        v1 = "\\begin{bmatrix} $a1 \\\\ $a2 \\end{bmatrix}"
        v2 = "\\begin{bmatrix} $b1 \\\\ $b2 \\end{bmatrix}"
        
        problem(
            question="Do the vectors \$\\mathbf{v}_1 = $v1\$ and \$\\mathbf{v}_2 = $v2\$ span \$\\mathbb{R}^2\$? Answer 'yes' or 'no'.",
            answer="no",
            difficulty=(1200, 1400),
            answer_type="choice",
            solution=steps(
                "Check if vectors are linearly independent",
                "Notice \$\\mathbf{v}_2 = $(k)\\mathbf{v}_1\$, so vectors are linearly dependent",
                "Two linearly dependent vectors cannot span \$\\mathbb{R}^2\$. Answer: no"
            ),
            time=90
        )
        
    elseif problem_type == 4
        # Dimension of subspace (1400-1600)
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)
        c = nonzero(-5, 5)
        
        problem(
            question="Find the dimension of the subspace \$W\$ of \$\\mathbb{R}^3\$ defined by \$W = \\{(x, y, z) : $(a)x + $(b)y + $(c)z = 0\\}\$.",
            answer=2,
            difficulty=(1400, 1600),
            solution=steps(
                "The equation \$$(a)x + $(b)y + $(c)z = 0\$ represents a plane through the origin",
                "This gives 1 constraint on 3 variables, leaving 2 free parameters",
                sol("Dimension", 2)
            ),
            time=100
        )
        
    elseif problem_type == 5
        # Coordinate vector relative to basis (1500-1700)
        # Basis vectors for R^2
        a1, a2 = nonzero(-5, 5), nonzero(-5, 5)
        b1, b2 = nonzero(-5, 5), nonzero(-5, 5)
        
        while a1 * b2 - a2 * b1 == 0
            b1, b2 = nonzero(-5, 5), nonzero(-5, 5)
        end
        
        # Target vector as combination
        c1, c2 = randint(-4, 4), randint(-4, 4)
        v1 = c1 * a1 + c2 * b1
        v2 = c1 * a2 + c2 * b2
        
        basis_v1 = "\\begin{bmatrix} $a1 \\\\ $a2 \\end{bmatrix}"
        basis_v2 = "\\begin{bmatrix} $b1 \\\\ $b2 \\end{bmatrix}"
        target_v = "\\begin{bmatrix} $v1 \\\\ $v2 \\end{bmatrix}"
        
        problem(
            question="Let \$\\mathcal{B} = \\{\\mathbf{b}_1, \\mathbf{b}_2\\}\$ where \$\\mathbf{b}_1 = $basis_v1\$ and \$\\mathbf{b}_2 = $basis_v2\$. Find the coordinate vector \$[\\mathbf{v}]_{\\mathcal{B}}\$ where \$\\mathbf{v} = $target_v\$. Enter your answer as an ordered pair (c₁, c₂).",
            answer=fmt_tuple((c1, c2)),
            difficulty=(1500, 1700),
            answer_type="tuple",
            solution=steps(
                "Solve \$\\mathbf{v} = c_1\\mathbf{b}_1 + c_2\\mathbf{b}_2\$",
                "This gives: \$$(v1) = $(a1)c_1 + $(b1)c_2\$ and \$$(v2) = $(a2)c_1 + $(b2)c_2\$",
                sol("[\\mathbf{v}]_{\\mathcal{B}}", fmt_tuple((c1, c2)))
            ),
            time=150
        )
        
    else
        # Nullspace dimension (1400-1700)
        # Create a matrix with known nullspace dimension
        rank_choice = rand(1:2)
        
        if rank_choice == 1
            # Rank 2 matrix in R^3 (nullspace dim = 1)
            a1, a2, a3 = nonzero(-5, 5), nonzero(-5, 5), nonzero(-5, 5)
            b1, b2, b3 = nonzero(-5, 5), nonzero(-5, 5), nonzero(-5, 5)
            
            # Ensure rows are independent
            while a1 * b2 == a2 * b1
                b1, b2 = nonzero(-5, 5), nonzero(-5, 5)
            end
            
            mat = "\\begin{bmatrix} $a1 & $a2 & $a3 \\\\ $b1 & $b2 & $b3 \\end{bmatrix}"
            ans = 1
        else
            # Rank 1 matrix in R^3 (nullspace dim = 2)
            a1, a2, a3 = nonzero(-5, 5), nonzero(-5, 5), nonzero(-5, 5)
            
            mat = "\\begin{bmatrix} $a1 & $a2 & $a3 \\end{bmatrix}"
            ans = 2
        end
        
        problem(
            question="Find the dimension of the nullspace of the matrix \$A = $mat\$.",
            answer=ans,
            difficulty=(1400, 1700),
            solution=steps(
                "Use the rank-nullity theorem: \$\\dim(\\text{null}(A)) + \\text{rank}(A) = n\$",
                "The matrix has $(rank_choice == 1 ? "2 rows, rank = 2" : "1 row, rank = 1"), and \$n = 3\$ columns",
                sol("\\dim(\\text{null}(A))", ans)
            ),
            time=120
        )
    end
end